src/decompressor.rs

```diff
@@ -0,0 +1,116 @@
+        let mut m = HashMap::new();
+        m.insert("gz", DecompressionCommand::new("gunzip", vec!["-c"]));
+        m.insert("bz2", DecompressionCommand::new("bunzip2", vec!["-c"]));
+        m.insert("xz", DecompressionCommand::new("unxz", vec!["-c"]));
+        m.insert("lzma", DecompressionCommand::new("unlzma", vec!["-c"]));
+        m
+    };
+    static ref SUPPORTED_COMPRESSION_FORMATS: GlobSet = {
+        let mut builder = GlobSetBuilder::new();
+        builder.add(Glob::new("*.gz").unwrap());
+        builder.add(Glob::new("*.bz2").unwrap());
+        builder.add(Glob::new("*.xz").unwrap());
+        builder.add(Glob::new("*.lzma").unwrap());
+        builder.build().unwrap()
+    };
+    static ref TAR_ARCHIVE_FORMATS: GlobSet = {
+        let mut builder = GlobSetBuilder::new();
+        builder.add(Glob::new("*.tar.gz").unwrap());
+        builder.add(Glob::new("*.tar.xz").unwrap());
+        builder.add(Glob::new("*.tar.bz2").unwrap());
+        builder.add(Glob::new("*.tgz").unwrap());
+        builder.add(Glob::new("*.txz").unwrap());
+        builder.add(Glob::new("*.tbz2").unwrap());
+        builder.build().unwrap()
+    };
+}
+
+/// Returns a handle to the stdout of the spawned decompression process for
+/// `path`, which can be directly searched in the worker.
+///
+/// If there is any error in spawning the decompression command, then return
+/// `None`, after outputting any necessary debug or error messages.
+pub fn get_reader(path: &Path, no_messages: bool) -> Option<ChildStdout> {
+    if is_tar_archive(path) {
+        debug!("Tar archives are currently unsupported: {}", path.display());
+        None
+    } else {
+        let extension = path.extension().and_then(OsStr::to_str).unwrap();
+        let decompression_command = DECOMPRESSION_COMMANDS.get(extension).unwrap();
+        let cmd = Command::new(decompression_command.cmd)
+            .args(decompression_command.args.as_slice())
+            .arg(path.as_os_str())
+            .stdout(Stdio::piped())
+            .stderr(Stdio::piped())
+            .spawn();
+        match cmd {
+            Ok(process) => {
+                let mut error = Vec::new();
+                // Unwrapping is safe here since we captured the stderr above
+                let mut stderr = process.stderr.unwrap();
+                // We drain the stderr of the child to ensure no error occurred,
+                // and then return the handle to stdout if that is the case
+                stderr.read_to_end(&mut error).unwrap();
+                if error.is_empty() {
+                    process.stdout
+                } else {
+                    if !no_messages {
+                        eprintln!("Error occurred while trying to decompress path: \
+                                   {}: {}", path.display(),
+                                  String::from_utf8_lossy(error.as_slice()));
+                    }
+                    None
+                }
```

Comment:
OK, so the issue here is that this will deadlock in non-trivial cases. You might not have hit it if you only tested on small files. Basically the issue here is that if you don't read from stdout, then the process you spawned (if it's implemented sanely) will at some point block on writing to stdout because it is connected to a pipe. If the calling process never reads from stdout, then the spawned process will eventually block. If the calling process is in turn blocked on something else from the spawned process---which in this case, we are, because we're trying to read from `stderr` until the spawned process stops writing to it---then you wind up with a deadlock.

If you only test this on small files then you won't notice this because the spawned process will likely be using buffered output. And since it's connected to a pipe, the buffer is usually biggish (~4K?) as opposed to being line buffered on a tty. As a result, the spawned process will finish doing its thing before ever flushing its buffer, which gives the appearance of code that works. :-)

The "fix" to this is to couple the `child` with reading stdout and wait to read stderr until the process terminates. This basically means that you can't return `ChildStdout` directly. You need to return something else that wraps up this logic.

Reply:
As an additional upside, the reader that you return can return normal error values, which are handled by the worker. This means you don't need to thread `no_messages` through the code.

Reply:
Ah, I knew it. 😞  When I made this change and looked at the code, I knew there had to be something I was missing.

Thank you for the detailed write up on the issues with my code. It makes a lot of sense to me and helps me understand it better. It also gives me incentive to do some more reading up on pipes and child processes, which I will. 👍